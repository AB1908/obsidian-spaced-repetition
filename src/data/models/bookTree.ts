import { annotation } from "src/data/import/annotations";
import { findPreviousHeader, Heading, isHeading } from "src/data/models/book";

// TODO: switch to DFS/BFS?
// Need this to be able to call countAnnotations

// TODO: why did I make this? Where do I need it?
// TODO: refactor
// TODO: think about heading collisions as there may be multiple chapters with same name
// TODO: don't need to nest paragraphs I think
// TODO: fix type

export function generateTree(sections: Heading[]) {
    let i = 0;
    let prevHeader;
    while (i < sections.length) {
        const cacheItem = sections[i];
        if (!isHeading(cacheItem)) {
            prevHeader = findPreviousHeader(cacheItem, sections);
            const previousHeader = sections[prevHeader] as Heading;
            if (prevHeader != null) {
                previousHeader.children.push(cacheItem);
            }
        }
        i++;
    }

    // successively attach lower level headers to higher ones
    // TODO: remove hardcoded headers and find the deepest header dynamically?
    // May be unnecessary since no book is going to have like a 5 level header... right??
    for (let headingLevel = 4; headingLevel > 1; headingLevel--) {
        let sectionIndex = 0;
        while (sectionIndex < sections.length) {
            const each = sections[sectionIndex];
            if ((isHeading(each)) && (each.level == headingLevel)) {
                const prevHeader = findPreviousHeader(each, sections);
                const previousHeader = sections[prevHeader] as Heading;
                previousHeader.children.push(each);
            }
            sectionIndex++;
        }
    }
    return sections.filter(t => isHeading(t) && t.level == 1);
}

export function generateSectionsTree(sections: (annotation | Heading)[]) {
    const headings: Heading[] = structuredClone(sections.filter((t): t is Heading => isHeading(t)));
    return generateTree(headings);
}