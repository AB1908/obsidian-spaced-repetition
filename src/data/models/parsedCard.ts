import {CardType} from "src/scheduling";
import {nanoid} from "nanoid";
import {writeCardToDisk} from "src/disk";
import {cardTextGenerator, generateCardAsStorageFormat, metadataTextGenerator} from "src/data/export/TextGenerator";
import {FLAG} from "src/data/deck";

export interface ParsedCard {
    id: string,
    cardType: CardType,
    cardText: string,
    metadataText: string,
    lineNo: number,
    notePath?: string,
}

// Note that this creates a learning card by default
export async function createParsedCard(
    questionText: string,
    answerText: string,
    cardType: CardType,
    path: string,
    annotationId: string,
    flag: FLAG = FLAG.LEARNING
): Promise<ParsedCard> {
    const parsedCard = {
        id: nanoid(8),
        notePath: path,
        cardText: cardTextGenerator(questionText, answerText, cardType),
        metadataText: metadataTextGenerator(annotationId, null, flag),
        // TODO: remove lineno
        lineNo: -1,
        cardType: cardType,
    };
    await writeCardToDisk(parsedCard.notePath, generateCardAsStorageFormat(parsedCard));
    return parsedCard;
}

export function createParsedCardFromText(
    cardText: string,
    cardType: CardType,
    path: string,
    metadataText: string,
): ParsedCard {
    return {
        id: nanoid(8),
        notePath: path,
        cardText: cardText,
        metadataText: metadataText,
        // TODO: remove lineno
        lineNo: -1,
        cardType: cardType,
    };
}