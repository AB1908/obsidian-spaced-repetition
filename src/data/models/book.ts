//todo: investigate using lowdb
import {getFileContents, getTFileForPath} from "src/disk";
import {annotation, parseAnnotations} from "src/data/import/annotations";
import {CachedMetadata, HeadingCache} from "obsidian";
import {nanoid} from "nanoid";
import {Flashcard} from "src/data/models/flashcard";
import {FlashCount} from "src/routes/notes-home-page";
import {generateFlashcardsArray, parseFileText} from "src/newparser";
import {ParsedCard} from "src/data/models/parsedCard";

// TODO: this is not really a "book" per se
export interface book {
    id:       string;
    name:     string;
    children: book[];
    counts?:  Counts;
}

export interface Counts {
    flashcards: FlashCount;
    sections: Record<string, Count>
}

export interface Count {
    with:    number;
    without: number;
}

// TODO: fix duplicate interface
export interface Annot {
    id: string;
    type: string;
    highlight: string;
    note: string;
    flashcards: Flashcard[]
}

function isHeading(section: annotation|Heading): section is Heading {
    return (section as Heading).level !== undefined;
}

function isAnnotation(section: annotation|Heading): section is annotation {
    return (section as annotation).highlight !== undefined;
}

// This is terrible. Save me.
function writeCountToObj(mem: any, sectionId: string, count: number, key: string) {
    Object.assign(mem, {[`${sectionId}`]: {...mem[sectionId], [key]: count}});
}

/*
This function is invoked twice to do a section tree walk, to figure out how many annotations have flashcards associated
with them and how many don't. I couldn't find a cleaner way of doing it without being too clever for myself.
 */
function countAnnotations(sections: any, mem: any, injectedCondition: (sections: any) => boolean, key: string) {
    let count = 0;
    if (isHeading(sections)) {
        for (let child of sections.children) {
            count += countAnnotations(child, mem, injectedCondition, key);
        }
        writeCountToObj(mem, sections.id, count, key);
    } else if (isAnnotation(sections)) {
        if (injectedCondition(sections)) {
            count += 1
        }
    }
    return count;
}

// TODO: switch to DFS/BFS?
export function AnnotationCount(sections: any) {
    let mem = {};
    // @ts-ignore
    mem[sections.id] = {
        "without": countAnnotations(sections, mem, (sections: any) => sections.hasFlashcards == false, "without"),
        "with": countAnnotations(sections, mem, (sections: any) => sections.hasFlashcards == true, "with")
    }
    return mem;
}

export async function createBook(path: string) {
    // some assumptions: one book note or multiple?
    const book = {
        path: path,
        contents: await getFileContents(path),
        // metadata: await getFileMetadata(path)
    }

    // get file contents
    // get file metadata
    // create array of sections
    // assign every section a uuid? This doesn't work. I need to get all annotations for a section
    // either by walking the tree or by finding all content up to the next section of the next level

    // create array of just paragraphs
    // create annotations array
    // push that into global annotation index
    //

    // i need a count of all annotations that are tested and not tested
    // i need section trees
    // let's put flashcard list here because I know it will be in the same folder
    // i should be able to get a list of annotations for any given section
    // i should be able to get a list of flashcards for any annotation
    // need to parse flashcards first so I know which annotations are covered and which are not
    // need a get next header function so i can find paragraphs belonging to current header
}

export function bookSections(metadata: CachedMetadata, fileText: string) {
    const output: (annotation|Heading)[] = [];
    const fileTextArray = fileText.split("\n");
    let headingIndex = 0;
    for (let cacheItem of metadata.sections) {
        // todo: consider parameterizing this
        if (cacheItem.type === "callout") {
            let annotation = parseAnnotations(fileTextArray.slice(cacheItem.position.start.line, cacheItem.position.end.line+1).join("\n"));
            output.push(annotation);
        } else if (cacheItem.type === "heading") {
            //done: fix casting
            output.push(new Heading(metadata.headings[headingIndex]));
            headingIndex++;
        } else {
            // TODO: Any edge cases?
        }
    }
    return output;
}

export class Heading {
    id: string;
    level: number;
    name: string;
    children: Heading[];

    constructor(heading: HeadingCache) {
        // might be too clever
        ({heading: this.name, level: this.level} = heading);
        this.id = nanoid(8);
        this.children = [];
    }

}

export type BookMetadataSections = (Heading | annotation)[];

export interface AnnotationWithFlashcard extends annotation {
    hasFlashcard: boolean;
}

// DONE rewrite to use ids instead of doing object equality
// DONE: fix types, narrowing doesn't work here somehow
export function getAnnotations(section: string, bookSections: BookMetadataSections) {
    const index = bookSections.findIndex(t => t.id === section);
    // todo: it feels like there should be a better way to do this
    const currentHeading = (bookSections[index] as Heading);
    let x = index + 1;
    while (x < bookSections.length) {
        const item = bookSections[x];
        // apparently, can't pass bookSections[x] in directly, see: https://stackoverflow.com/a/73666912
        if (isHeading(item)) {
            if (item.level <= currentHeading.level) {
                break;
            }
        }
        x++;
    }
    return bookSections.slice(index+1, x).filter(t => isAnnotation(t));
}

// Need this to be able to call countAnnotations
export function bookTree(id: string, name: string, bookSections: BookMetadataSections) {
    return {
        id,
        name,
        children: bookSections
    };
}

export interface frontbook {
    id:             string;
    name:           string;
    path:           string;
    parsedCards:    ParsedCard[];
    flashcards:     Flashcard[];
    annotationPath: string;
    // counts: BookCounts;
}

export interface BookCounts {
    mature:   number;
    new:      number;
    learning: number;
}

// todo: refactor to improve testability
export async function deckNote(path: string): Promise<frontbook> {
    const parsedCards = await parseFileText(path);
    return {
        id: nanoid(8),
        path: path,
        name: getTFileForPath(path).parent.name,
        parsedCards: parsedCards,
        flashcards: generateFlashcardsArray(parsedCards),
        annotationPath: "",
    };
}