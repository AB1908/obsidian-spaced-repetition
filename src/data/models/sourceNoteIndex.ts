import type SRPlugin from "src/main";
import { SourceNote } from "src/data/models/sourceNote";

export class SourceNoteIndex {
    sourceNotes: SourceNote[];

    constructor() {
        this.sourceNotes = [];
    }

    async initialize(plugin: SRPlugin) {
        // iterate over tags in plugin
        // create set from tags in note
        // check membership of tag
        const tagsInSettings = ["review/note", "review/book"];
        const pathsWithAllowedTags = new Set<string>();
        for (let [path, tags] of plugin.fileTagsMap) {
            const tagSet = new Set(tags);
            for (let tag of tagsInSettings) {
                if (tagSet.has(tag)) {
                    pathsWithAllowedTags.add(path);
                    break;
                }
            }
        }
        //todo: parameterize
        // const filePaths = filePathsWithTag("review/note");
        const notesWithAnnotations = Array.from(pathsWithAllowedTags.keys()).map((t: string) => new SourceNote(t, plugin));
        for (const t of notesWithAnnotations) {
            try {
                await t.initialize();
            } catch (e) {
                // WARNING! this is dangerous, I am catching other errors and just assuming that
                // these are this error
                console.error(e);
                console.error(`init: unable to initialize source note ${t.path}`);
            }
        }
        this.sourceNotes = notesWithAnnotations;
        console.log("Card Coverage: Source note index successfully initialized");
        return this;
    }

    getBook(bookId: string) {
        const book = this.sourceNotes.filter(t => t.id === bookId)[0];
        if (!book) {
            throw new Error(`No book found for id ${bookId}`);
        }
        return book;
    }

    getSourcesForReview() {
        return this.sourceNotes.filter(t => t.flashcardNote);
    }

    getSourcesWithoutFlashcards(): SourceNote[] {
        return this.sourceNotes.filter(t => !t.flashcardNote);
    }
}