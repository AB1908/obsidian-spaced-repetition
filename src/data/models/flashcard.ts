import { CardType, ReviewResponse, schedule } from "src/scheduler/scheduling";
import { nanoid } from "nanoid";
import { FLAG, FlashcardMetadata, parseCardText, parseFileText, parseMetadata } from "src/data/parser";
import { moment, Notice } from "obsidian";
import { SchedulingMetadata } from "src/data/utils/TextGenerator";
import { plugin } from "src/main";
import { ParsedCard } from "src/data/models/parsedCard";
import { getAnnotationFilePath } from "src/data/models/sourceNote";
import { filePathsWithTag, getCurrentDateIsoString } from "src/data/disk";
import { calculateDelayBeforeReview } from "src/utils";

export abstract class AbstractFlashcard {
    answerText: string;
    cardType: number;
    context: string;
    dueDate: string;
    ease: number;
    id: string;
    interval: number;
    questionText: string;
    siblings: string[];
    // todo: rename
    parentId: string;
    flag: FLAG;
    parsedCardId: string;

    // cardMetadata and highlight ID are mutually exclusive properties. Given that there is no constructor overloading
    // probably should change this to be a union type
    protected constructor(cardType: CardType, parsedCardId: string, cardMetadata?: FlashcardMetadata, annotationId?: string) {
        this.id = nanoid(8);
        this.cardType = cardType;
        this.context = null;
        this.dueDate = cardMetadata?.dueDate;
        this.ease = cardMetadata?.ease;
        this.interval = cardMetadata?.interval;
        this.parentId = cardMetadata?.annotationId || annotationId;
        this.flag = cardMetadata?.flag;
        this.siblings = [];
        this.parsedCardId = parsedCardId;
    }

    isDue() {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        return this.dueDate === null || this.dueDate <= getCurrentDateIsoString();
    }
}

export class Flashcard extends AbstractFlashcard {
    // cardMetadata and highlight ID are mutually exclusive properties. Given that there is no constructor overloading
    // probably should change this to be a union type
    constructor(parsedCardId: string, questionText: string, answerText: string, cardMetadata?: FlashcardMetadata, annotationId?: string) {
        // todo: handle other types later
        const cardType = CardType.MultiLineBasic;
        if (cardMetadata) {
            super(cardType, parsedCardId, cardMetadata);
        } else {
            super(cardType, parsedCardId, null, annotationId);
        }
        this.questionText = questionText;
        this.answerText = answerText;
    }
}

// todo: rename to update card?
export function schedulingMetadataForResponse(
    clickedResponse: ReviewResponse,
    schedulingMetadata: SchedulingMetadata,
): SchedulingMetadata {
    // const flashcard = getFlashcardById(flashcardId);
    // take the response received
    // use that to update flashcard internal state
    // that will take care of writing to disk
    // so this should be a relatively lean method
    // don't forget to update siblings?
    let schedObj;
    // is new card
    if (schedulingMetadata.dueDate === null) {
        // todo: move default settings down into schedule()?
        schedObj = schedule(
            clickedResponse,
            1.0,
            plugin.data.settings.baseEase,
            0,
        );
    } else {
        schedObj = schedule(
            clickedResponse,
            schedulingMetadata.interval,
            schedulingMetadata.ease,
            calculateDelayBeforeReview(schedulingMetadata.dueDate),
        );
    }
    const {interval, ease} = schedObj;
    // todo: parameterize format? nah
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const due = moment(Date.now() + interval * 24 * 3600 * 1000).format("YYYY-MM-DD");
    return {interval, ease, dueDate: due};
}

export function maturityCounts(flashcards: Flashcard[]) {
    let newCount = 0, mature = 0, learning = 0;
    flashcards.forEach(t => {
        if (t.interval >= 32) {
            mature += 1;
        } else if (t.interval && t.interval < 32) {
            learning += 1;
        } else {
            // } if (!t.interval) {
            newCount += 1;
        }
    });
    return {mature, learning, new: newCount};
}

export function createFlashcard(parsedCard: ParsedCard, questionText: string, answerText: string) {
    return new Flashcard(parsedCard.id, questionText, answerText, parseMetadata(parsedCard.metadataText));
}

export function generateFlashcardsArray(parsedCardsArray: ParsedCard[]) {
    const out: Flashcard[] = [];
    for (const parsedCard of parsedCardsArray) {
        const [questionText, answerText] = parseCardText(parsedCard.cardText);
        out.push(createFlashcard(parsedCard, questionText, answerText));
    }
    return out;
}

export class FlashcardNote {
    path: string;
    flashcards: Flashcard[];
    parentPath: string;
    parsedCards: ParsedCard[];

    constructor(path: string) {
        this.path = path;
        this.flashcards = [];
        this.parentPath = "";
        this.parsedCards = [];
    }

    async initialize(flashcardIndex: FlashcardIndex) {
        this.parentPath = getAnnotationFilePath(this.path);
        this.parsedCards = await parseFileText(this.path);
        try {
            this.flashcards = generateFlashcardsArray(this.parsedCards);
            this.flashcards.forEach(t => flashcardIndex.flashcards.set(t.id, t));
        } catch (e) {
            console.error(e);
            throw new Error(e);
        }
        return this;
    }
}

export class FlashcardIndex {
    flashcardNotes: FlashcardNote[];
    flashcards: Map<string, Flashcard>;

    constructor() {
        this.flashcardNotes = [];
        this.flashcards = new Map<string, Flashcard>();
    }

    async initialize() {
        // todo: remove hardcoding of tag
        const filePaths = filePathsWithTag("flashcards");
        const notesWithFlashcards = filePaths.map((t: string) => new FlashcardNote(t));
        for (const t of notesWithFlashcards) {
            try {
                await t.initialize(this);
            } catch (e) {
                // WARNING! this is dangerous, I am catching other errors and just assuming that these are this error
                console.error(e);
                console.error(`init: unable to initialize book ${t.path}`);
                new Notice("Error: Unable to parse legacy SRS flashcards. Try removing the #flashcards tag from files with SRS flashcards.");
            }
        }
        this.flashcardNotes = notesWithFlashcards;
        console.log("Card Coverage: Flashcard index successfully initialized");
        return this;
    }

    addFlashcardNoteToIndex(flashcardNote: FlashcardNote) {
        if (this.flashcardNotes.find(t => t.path === flashcardNote.path)) {
            throw new Error(`addFlashcardNoteToIndex: ${flashcardNote.path} is already in the index`);
        }
        this.flashcardNotes.push(flashcardNote);
    }
}