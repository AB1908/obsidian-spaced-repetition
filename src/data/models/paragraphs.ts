import { nanoid } from "nanoid";
import { getFileContents, getMetadataForFile } from "../disk";
import { Flashcard } from "./flashcard";
import { Heading } from "./book";

interface paragraph {
    id: string,
    text: string,
    wasIdPresent: boolean,
    hasFlashcards: boolean
}

// todo: edge case where there are no headings
export async function extractParagraphs(filePath: string, flashcards: Flashcard[]): Promise<paragraph[]> {
    const metadata = getMetadataForFile(filePath);
    // get blocks with blockid if present
    const paragraphsOrHeadings = metadata?.sections?.filter(t => t.type == "paragraph" || t.type == "heading");
    if (!paragraphsOrHeadings) {
        throw new Error(`getParagraphsFromFile: no paragraphs found in file ${filePath}`)
    }
    let headingIndex = 0;
    const headings = metadata?.headings;
    const fileContents = await getFileContents(filePath);
    const fileContentsArray = fileContents.split("\n");
    const processedData: paragraph[] = [];
    const paragraphsWithFlashcards = new Set(flashcards.map(t => t.annotationId));
    for (let section of paragraphsOrHeadings) {
        if (section.type == "paragraph") {
            const start = section.position.start.line;
            const end = section.position.end.line + 1;
            const paragraph = {
                id: section.id || nanoid(8),
                text: fileContentsArray.slice(start,end).join("\n"),
                wasIdPresent: section.id ? true : false,
            }
            processedData.push({
                ...paragraph,
                hasFlashcards: paragraphsWithFlashcards.has(paragraph.id), ...paragraph,
            })
        } else if (section.type == "heading") {
            // todo: this is an interesting type problem
            // whenever section.type is heading
            // we are guaranteed that headings is not null
            // therefore, we shouldn't have to worry about headings being undefined
            // and won't need to catch errors for that scenario
            // @ts-ignore
            processedData.push(new Heading(headings[headingIndex++]));
        }
    }
    return processedData;
    // still need header info
    // create interface from these attribs
}