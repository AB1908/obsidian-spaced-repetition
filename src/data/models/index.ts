import { SourceNote } from "src/data/models/sourceNote";
import { FlashcardNote } from "src/data/models/flashcard";
import { annotation } from "src/data/models/annotations";
import { paragraph } from "src/data/models/paragraphs";

export class Index {
    sourceNoteIndex: Map<string, SourceNote>;
    flashcardNoteIndex: Map<string, FlashcardNote>;
    annotationIndex: Map<string, (paragraph|annotation)>

    constructor() {
        this.annotationIndex = new Map<string, paragraph | annotation>();
    }

    initialize() {

    }

    addToAnnotationIndex(block: annotation|paragraph) {
        // todo: think about incorporating an index dirty state to
        // properly retrigger indexing
        this.annotationIndex.set(block.id, block);
    }

    getAnnotation(blockId: string) {
        return this.annotationIndex.get(blockId);
    }
}