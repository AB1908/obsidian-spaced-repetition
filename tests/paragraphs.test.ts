import type { Flashcard } from "src/data/models/flashcard";
import { extractParagraphs } from "src/data/models/paragraphs";

jest.mock("../src/main");

// eslint-disable-next-line @typescript-eslint/no-empty-function
jest.mock("../src/data/disk", () => {
    return {
        getMetadataForFile: (path: string) => {
            return {
                "headings": [
                    {
                        "position": {
                            "start": {
                                "line": 0,
                                "col": 0,
                                "offset": 0
                            },
                            "end": {
                                "line": 0,
                                "col": 46,
                                "offset": 46
                            }
                        },
                        "heading": "Chapter 3: Pulling the rabbit out of the hat",
                        "level": 1
                    },
                    {
                        "position": {
                            "start": {
                                "line": 2,
                                "col": 0,
                                "offset": 48
                            },
                            "end": {
                                "line": 2,
                                "col": 26,
                                "offset": 74
                            }
                        },
                        "heading": "Relating study and test",
                        "level": 2
                    }
                ],
                "sections": [
                    {
                        "type": "heading",
                        "position": {
                            "start": {
                                "line": 0,
                                "col": 0,
                                "offset": 0
                            },
                            "end": {
                                "line": 0,
                                "col": 46,
                                "offset": 46
                            }
                        }
                    },
                    {
                        "type": "heading",
                        "position": {
                            "start": {
                                "line": 2,
                                "col": 0,
                                "offset": 48
                            },
                            "end": {
                                "line": 2,
                                "col": 26,
                                "offset": 74
                            }
                        }
                    },
                    {
                        "type": "paragraph",
                        "position": {
                            "start": {
                                "line": 4,
                                "col": 0,
                                "offset": 76
                            },
                            "end": {
                                "line": 5,
                                "col": 23,
                                "offset": 124
                            }
                        },
                        "id": "hjhjhlkap"
                    }
                ],
                "blocks": {
                    "hjhjhlkap": {
                        "position": {
                            "start": {
                                "line": 4,
                                "col": 0,
                                "offset": 76
                            },
                            "end": {
                                "line": 5,
                                "col": 23,
                                "offset": 124
                            }
                        },
                        "id": "hjhjhlkap"
                    }
                }
            };
        },
        getFileContents: (path: string) => {
            return `# Chapter 3: Pulling the rabbit out of the hat

## Relating study and test

What is episodic memory?
NO one knows ^hjhjhlkap`;
        }
    };
});

describe("extractParagraphs", () => {
    beforeEach(async () => {
        // todo: fix ts error
        const nanoid = require("nanoid");
        let value = 0;
        nanoid.nanoid.mockImplementation((_size?: number) => (value++).toString());
    });
    test("should extract an array of headers from text", async () => {
        const flashcards = [
            {
                "annotationId": "93813",
                "answerText": "This is an answer",
                "cardType": 2,
                "context": null,
                "dueDate": null,
                "ease": null,
                "flag": null,
                "id": "b4cYyMuF",
                "interval": null,
                "parsedCardId": "aaaaaaaa",
                "questionText": "This is a question",
                "siblings": []
            }
        ];
        expect(await extractParagraphs("", flashcards as unknown as Flashcard[])).toMatchSnapshot();
        flashcards[0].annotationId = "hjhjhlkap";
        expect(await extractParagraphs("", flashcards as unknown as Flashcard[])).toMatchSnapshot();
    });
    afterEach(() => {
        jest.restoreAllMocks();
    });
});