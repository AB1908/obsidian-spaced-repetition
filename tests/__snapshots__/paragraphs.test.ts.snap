// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getParagraphFromFile 1`] = `
[
  Heading {
    "children": [],
    "counts": {
      "with": 0,
      "without": 0,
    },
    "id": "0",
    "level": 1,
    "name": "Chapter 3: Pulling the rabbit out of the hat",
  },
  Heading {
    "children": [],
    "counts": {
      "with": 0,
      "without": 0,
    },
    "id": "1",
    "level": 2,
    "name": "Relating study and test",
  },
  {
    "hasFlashcards": false,
    "id": "hjhjhlkap",
    "text": "What is episodic memory?
NO one knows ^hjhjhlkap",
    "wasIdPresent": true,
  },
]
`;

exports[`getParagraphFromFile 2`] = `
[
  Heading {
    "children": [],
    "counts": {
      "with": 0,
      "without": 0,
    },
    "id": "2",
    "level": 1,
    "name": "Chapter 3: Pulling the rabbit out of the hat",
  },
  Heading {
    "children": [],
    "counts": {
      "with": 0,
      "without": 0,
    },
    "id": "3",
    "level": 2,
    "name": "Relating study and test",
  },
  {
    "hasFlashcards": true,
    "id": "hjhjhlkap",
    "text": "What is episodic memory?
NO one knows ^hjhjhlkap",
    "wasIdPresent": true,
  },
]
`;
